cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCredentials.scala

```diff
@@ -0,0 +1,59 @@
+package cloud.nio.impl.drs
+
+import cats.syntax.validated._
+import com.azure.identity.DefaultAzureCredentialBuilder
+import com.azure.security.keyvault.secrets.{SecretClient, SecretClientBuilder}
+import com.google.auth.oauth2.{AccessToken, OAuth2Credentials}
+import common.validation.ErrorOr
+import common.validation.ErrorOr.ErrorOr
+
+import scala.concurrent.duration._
+
+/**
+  * This trait allows us to abstract away different token attainment strategies
+  * for different cloud environments.
+  **/
+sealed trait DrsCredentials {
+  def getAccessToken(acceptableTTL: Duration): ErrorOr[String]
+}
+
+case class GoogleDrsCredentials(credentials: OAuth2Credentials) extends DrsCredentials {
+  //Based on method from GoogleRegistry
+  def getAccessToken(acceptableTTL: Duration): ErrorOr[String] = {
+    def accessTokenTTLIsAcceptable(accessToken: AccessToken): Boolean = {
+      (accessToken.getExpirationTime.getTime - System.currentTimeMillis()).millis.gteq(acceptableTTL)
+    }
+
+    Option(credentials.getAccessToken) match {
+      case Some(accessToken) if accessTokenTTLIsAcceptable(accessToken) =>
+        accessToken.getTokenValue.validNel
+      case _ =>
+        credentials.refresh()
+        Option(credentials.getAccessToken.getTokenValue) match {
+          case Some(accessToken) => accessToken.validNel
+          case None => "Could not refresh access token".invalidNel
+        }
+    }
+  }
+}
+
+case class AzureDrsCredentials(identityClientId: Option[String], vaultName: String, secretName: String) extends DrsCredentials {
+
+  lazy val secretClient: ErrorOr[SecretClient] = ErrorOr {
+    val defaultCreds = identityClientId.map(identityId =>
+      new DefaultAzureCredentialBuilder().managedIdentityClientId(identityId)
+    ).getOrElse(
+      new DefaultAzureCredentialBuilder()
+    ).build()
+
+    new SecretClientBuilder()
+      .vaultUrl(s"https://${vaultName}.vault.azure.net")
+      .credential(defaultCreds)
+      .buildClient()
+  }
+
+  def getAccessToken(acceptableTTL: Duration): ErrorOr[String] = {
+    // TTL value is unused
```

Comment:
Should acceptable TTL be rolled into `GoogleDrsCredentials`? I'm guessing that might be a non-trivial refactor.

Reply:
Huh, my initial thought was that we couldn't because we didn't have access to the TTL when `GoogleDrsCredentials` is created, but I think it might be easier than I thought. I'll give it a try.

Reply:
Made this change, let me know what you think. I like it better this way, though it took a fair amount of test-munging.